## Data Universes

1. Data is a finite sequence of codas.
2. A coda is a pair of data. 

This defines and countably infinite universe of data.  We can consider cutting these off at some maximum depth or width.  This could be consistently implemented by defining the "outer crust" to be atomic or by mapping them all to empty data.  All sorts of interesting questions arise.

In [1]:
import itertools
import Universe

def f(depth,width): 
    U = Universe.alldata(depth,width)
    print(depth,width,'->',len(U),len(set(U)))

In [2]:
f(5,1)

5 1 -> 458330 458330


### Universe sizes width (horizontal) vs depth (vertical)

| width:   |    1    |    2   |    3   |    4    |    5     |
|---------:|--------:|-------:|-------:|--------:|---------:|
| depth  1      |    2    |   3    |   4     |    5     |  6  |
| depth  2      |    5    |  91    | 4369   | 406901  |    |
| depth  3      |   26    | 68583243     |        |         |   |
| depth  4      |  677    |        |        |         |    |
| depth  5      | 458330  |        |        |         |    |

 

In [7]:
n = 0 
for d in Universe.alldata(2,2): 
    n += 1 
    print(n,'<-',repr(d),len(d),d.depth())

1 <-  0 0
2 <- (:) 1 1
3 <- (:(:)) 1 2
4 <- (:(:)(:)) 1 2
5 <- ((:):) 1 2
6 <- ((:):(:)) 1 2
7 <- ((:):(:)(:)) 1 2
8 <- ((:)(:):) 1 2
9 <- ((:)(:):(:)) 1 2
10 <- ((:)(:):(:)(:)) 1 2
11 <- (:)(:) 2 1
12 <- (:)(:(:)) 2 2
13 <- (:)(:(:)(:)) 2 2
14 <- (:)((:):) 2 2
15 <- (:)((:):(:)) 2 2
16 <- (:)((:):(:)(:)) 2 2
17 <- (:)((:)(:):) 2 2
18 <- (:)((:)(:):(:)) 2 2
19 <- (:)((:)(:):(:)(:)) 2 2
20 <- (:(:))(:) 2 2
21 <- (:(:))(:(:)) 2 2
22 <- (:(:))(:(:)(:)) 2 2
23 <- (:(:))((:):) 2 2
24 <- (:(:))((:):(:)) 2 2
25 <- (:(:))((:):(:)(:)) 2 2
26 <- (:(:))((:)(:):) 2 2
27 <- (:(:))((:)(:):(:)) 2 2
28 <- (:(:))((:)(:):(:)(:)) 2 2
29 <- (:(:)(:))(:) 2 2
30 <- (:(:)(:))(:(:)) 2 2
31 <- (:(:)(:))(:(:)(:)) 2 2
32 <- (:(:)(:))((:):) 2 2
33 <- (:(:)(:))((:):(:)) 2 2
34 <- (:(:)(:))((:):(:)(:)) 2 2
35 <- (:(:)(:))((:)(:):) 2 2
36 <- (:(:)(:))((:)(:):(:)) 2 2
37 <- (:(:)(:))((:)(:):(:)(:)) 2 2
38 <- ((:):)(:) 2 2
39 <- ((:):)(:(:)) 2 2
40 <- ((:):)(:(:)(:)) 2 2
41 <- ((:):)((:):) 2 2
42 <- ((:):)((:):(:)) 2 